### PR TITLE
feat: create graph structure

### PR DESCRIPTION
**Pull request by GabinDDL**
Add graph structure with utilities functions : 
- get_neighbors
- pretty_printers
- condition functions
- search functions
- init function for toroidal grid (it works with array because of complexity)

Due to lack of time, function tests have only been hand-tested, and further tests will be added in a future MR.

### CODE REVIEW

File: lib/graph.ml [hunk 1]
```diff
@@ -0,0 +1,90 @@
+type color = Red | Blue | Green | Yellow
```
Comment (ThomArr) [inline lib/graph.ml:1]: I thought I would put an int. As the Mr Rabie had shown in his examples.

File: lib/graph.ml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+type position = int * int
+type vertex = position * color
+type neighbors = vertex list
+type graph = int * (vertex * neighbors) list
+type graph_traversal_error = UnknownVertex
```
Comment (ThomArr) [inline lib/graph.ml:6]: Would you say exception?

File: lib/graph.ml [hunk 3]
```diff
@@ -0,0 +1,90 @@
+type color = Red | Blue | Green | Yellow
+type position = int * int
```
Comment (ThomArr) [inline lib/graph.ml:2]: Maybe give them names. Even though I find this representation very odd what do you mean by int Int. The row and the column? But what is the first row and the first column in a torus:/.
I personally thought more about having a list of neighbors. Maybe discuss it with Mr. Rabie?

File: lib/graph.ml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+type vertex = position * color
+type neighbors = vertex list
+type graph = int * (vertex * neighbors) list
```
Comment (ThomArr) [inline lib/graph.ml:5]: Maybe give names to your variables. What is int?

File: lib/graph.ml (hunk 1 above)
Comment (GabinDDL) [reply lib/graph.ml:1]: It's mainly to better differentiate them from the coordinates, they're always comparable too, that's why I made this choice, after it's true that it restricts the colors to a finite number, I'd surely make the change

File: lib/graph.ml (hunk 2 above)
Comment (GabinDDL) [reply lib/graph.ml:6]: No it's the error types for the result

File: lib/graph.ml (hunk 4 above)
Comment (GabinDDL) [reply lib/graph.ml:5]: it's for the time, but it's true that a structure might be better, I'll modify it.

File: lib/graph.ml (hunk 3 above)
Comment (GabinDDL) [reply lib/graph.ml:2]: for positions, this is how it's normally represented: x, y with x the column number and y the row number, it's common in programs, so I don't see the point of specifying it

**Comment by ThomArr** [comment]
Sorry, but the main problem is that I’m not surprised at the way you modeled the thing.  I thought a knot would have a color and a list of neighbors. And a graph would have been a time and a root. Should we discuss this with Mr Rabie to see what would be best?

**Comment by GabinDDL** [comment]
> Sorry, but the main problem is that I’m not surprised at the way you modeled the thing. I thought a knot would have a color and a list of neighbors. And a graph would have been a time and a root. Should we discuss this with Mr Rabie to see what would be best?

Personally, I've adopted the basic structure of a graph (which doesn't admit a root but a set of vertices and neighbors that is an adjacency list), adding an int attribute for time (I'll probably modify it so that it has a name, adding the dimensions too), which is why I'm specifying the coordinates inside to be able to differentiate them within a grid (like just giving them a name). You can't differentiate them with just colors, and a path is then no longer possible, hence my choice. If you have any doubts, we can discuss them with Mr Rabbie, yes.

**Comment by GabinDDL** [comment]
@ThomArr I've resolved what I think should be modified about what you asked

**Review by ThomArr — APPROVED** [review]

**Pull request merged**